noise_corrector.py

```diff
@@ -71,149 +216,238 @@ def estimate_true_values(observed_values, randomized_response_rate, beta=0):
-  # 2925.
-  # https://en.wikipedia.org/wiki/Stars_and_bars_(combinatorics)
-  assert len(observed_values) == 2925
-  return estimate_true_values(observed_values,
-                              RANDOMIZED_RESPONSE_RATE_NAV_SOURCES)
-
-
-def simulate_randomization(true_reports, randomized_response_rate):
-  """Simulates randomized response on true_reports; this simulates the way the
-  browser would treat reports in an idealized system (e.g. no rate limits hit,
-  etc)."""
-  n = sum(true_reports)
-  k = len(true_reports)
-  noisy_reports = [0] * k
-
-  # This method could be implemented by doing randomized response on every
-  # individual true report. However, we can optimize this by directly computing
-  # the number of total responses that would have flipped, and uniformly
-  # distributing them across all k buckets.
-  x = randomized_response_rate
-
-  # Compute the non-flipped counts.
-  non_flipped_counts = [numpy.random.binomial(true_count, 1 - x) for true_count in true_reports]
-
-  # Distribute the flipped reports uniformly among all k buckets.
-  num_flipped = n - sum(non_flipped_counts)
-  flipped_counts = numpy.random.multinomial(num_flipped, [1/k] * k)
-  return non_flipped_counts + flipped_counts
-
-if __name__ == "__main__":
-  run_example()
+    return (v - n * x / k) / (1 - x)
+
+  return {k: estimate(v) for k, v in observed_values.items()}
+
+
+def join_reports_with_sources(sources: List[Source], reports: List[Report]) -> JoinedGen:
+  '''Returns tuple of (source, [reports]) joined with the source_event_id'''
+  report_map = collections.defaultdict(list)
+  for r in reports:
+    report_map[r['report']['source_event_id']].append(r)
+
+  def join(source) -> Joined:
+    source_event_id = source['registration_config']['source_event_id']
+    return (source, report_map[source_event_id])
+  return (join(s) for s in sources)
+
+
+################## AGGREGATE UTILITIES ##################
+
+def get_raw_corrected_map(joined: Iterable[Joined], params: ParamConfig) -> Dict[OutputEnumeration, float]:
+  '''Returns an aggregate map of OutputEnumeration -> float, with noise debiased
+  according to params'''
+  output_counts = {o: 0 for o in OutputEnumeration.generate_all(params)}
+  for j in joined:
+    output = OutputEnumeration.create_from_data(j)
+    assert output in output_counts
+    output_counts[output] += 1
+
+  return estimate_true_values(output_counts, params.rr_rate)
+
+
+def correct_aggregates(joined: JoinedGen, params: ParamConfig) -> List[dict]:
+  '''Returns debiased report counts broken out by trigger_data'''
+  corrected = get_raw_corrected_map(joined, params)
+  histogram = numpy.zeros(params.data_cardinality)
+  for o, c in corrected.items():
+    histogram += c * o.data_histogram()
+  x = 10
+  return [{'trigger_data': t,
+           'report_count': v} for t, v in enumerate(histogram)]
+
+
+################## EVENT-LEVEL UTILITIES ##################
+
+def adjust_to_match_distribution(values: List[T],
```

Comment:
Is `adjust_to_match_distribution` something we expect devs to use directly?
I'm confused with the concept of aggregate distribution in the context of event-level reports.

Reply:
No, we don't expect them to use this directly.  I tried to explain this more in the help text

Reply:
Nice, I think it's helpful 👍